fru.c

```diff
@@ -280,30 +284,33 @@ fru_field_t * fru_encode_data(int len, const uint8_t *data)
 }
 
 /**
- * Allocate a buffer and decode the data from it.
+ * Decode data from a buffer into another buffer.
+ *
+ * For binary data use FRU_FIELDDATALEN(typelen) to find
+ * out the size of valid bytes in the returned buffer.
  *
- * For binary data use FRU_FIELDDATALEN(field->typelen) to find
- * out the size of the returned buffer.
+ * Return false if there were errors during decoding and true otherwise.
  */
-unsigned char * fru_decode_data(const fru_field_t *field)
+bool fru_decode_data(uint8_t typelen, //< [in] typelen for the raw data
+                     const uint8_t *data, //<[in] buffer containing raw data
```

Comment:
Since the input to the decode functions I've written are `fru_X_area_t` they each already have a data buffer, but if I wanted a `fru_field_t`, I'd need to allocate a buffer and copy the data. Seemed wasteful.

Reply:
I could use `fru_field_t` if I changed its definition to be typelen + ptr to data instead of storing the data adjacent to the struct.

Reply:
With typelen+ptr it wouldn't be suitable for type punning anymore. But you for some reason don't use type punning. Instead of this:

```
chassis_out->type = area->langtype;
const uint8_t *curr_data = area->data;
typelen = curr_data[0];
curr_data++;

if (!fru_decode_data(typelen, curr_data, chassis_out->pn,
                                sizeof(chassis_out->pn))
{
 ...
}
```

You could use
```
chassis_out->type = area->langtype;
fru_field_t *field = area->data;

if (!fru_decode_data(field, chassis_out->pn, sizeof(chassis_out->pn)) {
 ...
}
```
and have the same result, yet in a more elegant way.

I agree that the version of `fru_decode_data()` that takes a pointer to an already allocated buffer may be more useful than the original version with internal allocation, but I'm not convinced that ditching `fru_field_t` helps in any way.